In [67]:
# Importing the required libraries
import pandas as pd

In [68]:
# load datasets
rating = pd.read_csv("ratings.csv")
rating.head(2)

userId  movieId  rating   timestamp
0       1       16     4.0  1217897793
1       1       24     1.5  1217895807

In [69]:
print("The Shape is:")
rating.shape

The Shape is:


(105339, 4)

In [70]:
movies = pd.read_csv("movies.csv")
movies = movies.iloc[:, 0:2]
movies.head(2)

movieId             title
0        1  Toy Story (1995)
1        2    Jumanji (1995)

In [71]:
print("The Shape is:")
movies.shape

The Shape is:


(10329, 2)

In [72]:
# Merge
df = pd.merge(rating, movies, on='movieId')
df.head(10)

userId  movieId  rating   timestamp          title
0       1       16     4.0  1217897793  Casino (1995)
1       9       16     4.0   842686699  Casino (1995)
2      12       16     1.5  1144396284  Casino (1995)
3      24       16     4.0   963468757  Casino (1995)
4      29       16     3.0   836820223  Casino (1995)
5      31       16     4.0   832397760  Casino (1995)
6      47       16     4.0   853361808  Casino (1995)
7      60       16     4.0  1162941904  Casino (1995)
8      62       16     4.5  1267576909  Casino (1995)
9      88       16     3.5  1180879072  Casino (1995)

In [73]:
print("The Shape is:")
df.shape

The Shape is:


(105339, 5)

In [74]:
# We won't use the timestamp, let's drop it
df.drop(["timestamp"], axis=1, inplace=True)
df.head(7)

userId  movieId  rating          title
0       1       16     4.0  Casino (1995)
1       9       16     4.0  Casino (1995)
2      12       16     1.5  Casino (1995)
3      24       16     4.0  Casino (1995)
4      29       16     3.0  Casino (1995)
5      31       16     4.0  Casino (1995)
6      47       16     4.0  Casino (1995)

In [75]:
# Set Up Our Recommendation System
moviemat = df.pivot_table(index='userId', columns='title', values='rating')
moviemat.head(4)

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              NaN                                      NaN   
2              NaN                                      NaN   
3              NaN                                      NaN   
4              NaN                                      NaN   

title   'Round Midnight (1986)  'Til There Was You (1997)  'burbs, The (1989)  \
userId                                                                          
1                          NaN                        NaN                 NaN   
2                          NaN                        NaN                 NaN   
3                          NaN                        NaN                 NaN   
4                          NaN                        NaN                 NaN   

title   'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                      
1                        NaN                          NaN   
2                        NaN                          NaN   
3                        NaN                          NaN   
4                        NaN                          NaN   

title   *batteries not included (1987)  ...And Justice for All (1979)  \
userId                                                                  
1                                  NaN                            NaN   
2                                  NaN                            NaN   
3                                  NaN                            NaN   
4                                  NaN                            NaN   

title   10 (1979)  ...  [REC] (2007)  [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId             ...                                                         
1             NaN  ...           NaN            NaN                      NaN   
2             NaN  ...           NaN            NaN                      NaN   
3             NaN  ...           NaN            NaN                      NaN   
4             NaN  ...           NaN            NaN                      NaN   

title   a/k/a Tommy Chong (2005)  eXistenZ (1999)  \
userId                                              
1                            NaN              NaN   
2                            NaN              NaN   
3                            NaN              NaN   
4                            NaN              NaN   

title   loudQUIETloud: A Film About the Pixies (2006)  xXx (2002)  \
userId                                                              
1                                                 NaN         NaN   
2                                                 NaN         NaN   
3                                                 NaN         NaN   
4                                                 NaN         NaN   

title   xXx: State of the Union (2005)  ¡Three Amigos! (1986)  \
userId                                                          
1                                  NaN                    NaN   
2                                  NaN                    NaN   
3                                  NaN                    NaN   
4                                  NaN                    NaN   

title   À nous la liberté (Freedom for Us) (1931)  
userId                                             
1                                             NaN  
2                                             NaN  
3                                             NaN  
4                                             NaN  

[4 rows x 10323 columns]

In [76]:
print("The type is:")
type(moviemat)

The type is:


pandas.core.frame.DataFrame

In [77]:
# Let's look at the user ratings of the film Braveheart (1995):
Braveheart_user_ratings = moviemat["Braveheart (1995)"]
Braveheart_user_ratings.head()

userId
1    4.0
2    NaN
3    NaN
4    NaN
5    0.5
Name: Braveheart (1995), dtype: float64

In [78]:
# Let's calculate the correlation of all films
# with the Braveheart film using the corrwith() method:
similar_to_Braveheart = moviemat.corrwith(Braveheart_user_ratings)
similar_to_Braveheart.head(30)
#It gives the following error due to NaN values.

c:\Python311\Lib\site-packages\numpy\lib\function_base.py:2889: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar, dtype=dtype)
c:\Python311\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: divide by zero encountered in divide
  c *= np.true_divide(1, fact)
c:\Python311\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning: invalid value encountered in multiply
  c *= np.true_divide(1, fact)
c:\Python311\Lib\site-packages\numpy\lib\function_base.py:2897: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
c:\Python311\Lib\site-packages\numpy\lib\function_base.py:2898: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]


title
'71 (2014)                                                     NaN
'Hellboy': The Seeds of Creation (2004)                        NaN
'Round Midnight (1986)                                         NaN
'Til There Was You (1997)                                      NaN
'burbs, The (1989)                                        0.660184
'night Mother (1986)                                           NaN
(500) Days of Summer (2009)                              -0.012659
*batteries not included (1987)                            0.058282
...And Justice for All (1979)                            -0.308957
10 (1979)                                                 1.000000
10 Items or Less (2006)                                  -1.000000
10 Things I Hate About You (1999)                         0.557778
10 to Midnight (1983)                                          NaN
10,000 BC (2008)                                         -0.687500
100 Girls (2000)                                        


According to this result, the correlation between the Braveheart (1995) film  and the film
"burbs, The (1989)" is 0.660184. In other words, 66% of the people who watched
the Braveheart (1995) film also watched the film "burbs, The (1989)".

On the other hand, the correlation between the film "Braveheart (1995)" and the film "102 Dalmatians (2000)"
was -0.970725. In other words, 97% of the people who watched the Braveheart  film did not watch
 the film "102 Dalmatians (2000)".

In [79]:
print("The type is:")
type(similar_to_Braveheart)

The type is:


pandas.core.series.Series

In [80]:
# The list we've got is a Pandas series. If we convert it to df, we can work more easily.
corr_Braveheart = pd.DataFrame(similar_to_Braveheart, columns=['Correlation'])
corr_Braveheart.dropna(inplace=True)
corr_Braveheart.head()

Correlation
title                                      
'burbs, The (1989)                 0.660184
(500) Days of Summer (2009)       -0.012659
*batteries not included (1987)     0.058282
...And Justice for All (1979)     -0.308957
10 (1979)                          1.000000

In [81]:
# Let's list the data frame we obtained and see which film it suggests as the closest to Braveheart
corr_Braveheart.sort_values('Correlation', ascending=False).head(10)

Correlation
title                                             
Bigger, Stronger, Faster* (2008)               1.0
Delta Farce (2007)                             1.0
Enemy (2013)                                   1.0
Avalanche (1978)                               1.0
The Alphabet Killer (2008)                     1.0
Bon Cop, Bad Cop (2006)                        1.0
Texas Chainsaw Massacre 2, The (1986)          1.0
Loverboy (1989)                                1.0
All the King's Men (2006)                      1.0
Main Event, The (1979)                         1.0

In [82]:
# Let's find the average (mean value) rating of each film
ratings = pd.DataFrame(df.groupby('title')['rating'].mean())
ratings

rating
title                                                  
'71 (2014)                                     3.500000
'Hellboy': The Seeds of Creation (2004)        3.000000
'Round Midnight (1986)                         2.500000
'Til There Was You (1997)                      4.000000
'burbs, The (1989)                             3.125000
...                                                 ...
loudQUIETloud: A Film About the Pixies (2006)  4.500000
xXx (2002)                                     2.958333
xXx: State of the Union (2005)                 2.071429
¡Three Amigos! (1986)                          3.012500
À nous la liberté (Freedom for Us) (1931)      3.000000

[10323 rows x 1 columns]

In [83]:
# Let's sort it
ratings.sort_values('rating', ascending=False).head(20)

rating
title                                                     
Saddest Music in the World, The (2003)                 5.0
Interstate 60 (2002)                                   5.0
Gunfighter, The (1950)                                 5.0
Heima (2007)                                           5.0
Limelight (1952)                                       5.0
Plague Dogs, The (1982)                                5.0
Love Me If You Dare (Jeux d'enfants) (2003)            5.0
Syrup (2013)                                           5.0
Interstella 5555: The 5tory of the 5ecret 5tar ...     5.0
Symbol (Shinboru) (2009)                               5.0
Friend Is a Treasure, A (Chi Trova Un Amico, Tr...     5.0
Memories (Memorîzu) (1995)                             5.0
50 Children: The Rescue Mission of Mr. And Mrs....     5.0
Olive Kitteridge (2014)                                5.0
Switchblade Sisters (1975)                             5.0
Forbidden Zone (1980)                                  5.0
Possession (1981)                                      5.0
Swan, The (1956)                                       5.0
Lawless Heart, The (2003)                              5.0
Enigma of Kaspar Hauser, The (a.k.a. Mystery of...     5.0

In [84]:
ratings['rating_vote_number'] = pd.DataFrame(df.groupby('title')['rating'].count())
ratings.head()

rating  rating_vote_number
title                                                              
'71 (2014)                                3.500                   1
'Hellboy': The Seeds of Creation (2004)   3.000                   1
'Round Midnight (1986)                    2.500                   1
'Til There Was You (1997)                 4.000                   3
'burbs, The (1989)                        3.125                  20

In [85]:
# Now let's take a look at the most voted films in descending order¶
ratings.sort_values('rating_vote_number', ascending=False).head()

rating  rating_vote_number
title                                                         
Pulp Fiction (1994)               4.160000                 325
Forrest Gump (1994)               4.138264                 311
Shawshank Redemption, The (1994)  4.454545                 308
Jurassic Park (1993)              3.659864                 294
Silence of the Lambs, The (1991)  4.194828                 290

In [86]:
# add the rating_vote_number column to our corr_Braveheart dataframe (with join)
corr_Braveheart.sort_values('Correlation', ascending=False).head(10)

Correlation
title                                             
Bigger, Stronger, Faster* (2008)               1.0
Delta Farce (2007)                             1.0
Enemy (2013)                                   1.0
Avalanche (1978)                               1.0
The Alphabet Killer (2008)                     1.0
Bon Cop, Bad Cop (2006)                        1.0
Texas Chainsaw Massacre 2, The (1986)          1.0
Loverboy (1989)                                1.0
All the King's Men (2006)                      1.0
Main Event, The (1979)                         1.0

In [87]:
corr_Braveheart = corr_Braveheart.join(ratings['rating_vote_number'])
corr_Braveheart.head()

Correlation  rating_vote_number
title                                                          
'burbs, The (1989)                 0.660184                  20
(500) Days of Summer (2009)       -0.012659                  37
*batteries not included (1987)     0.058282                  11
...And Justice for All (1979)     -0.308957                  10
10 (1979)                          1.000000                   3

NOTE: Films with a small number of votes can mislead us.
For example, the fact that 1-2 people gave a film 5 does not mean the same thing as the fact that
100 people gave it 5. For this reason, it would be much more accurate to decide by looking at
the average ratings of films with more than 100 votes

In [88]:
final = corr_Braveheart[corr_Braveheart['rating_vote_number']> 100].sort_values('Correlation', ascending=False)
final.head()


Correlation  rating_vote_number
title                                                      
Braveheart (1995)              1.000000                 248
Jerry Maguire (1996)           0.471669                 124
Green Mile, The (1999)         0.440321                 101
American History X (1998)      0.439146                 103
In the Line of Fire (1993)     0.430056                 105

In [89]:
final.index[1]

'Jerry Maguire (1996)'

In [90]:
print("Final Recommodition: You can also watch the ************",final.index[1], "****************")

Final Recommodition: You can also watch the ************ Jerry Maguire (1996) ****************
